Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -24,28 +26,85 @@ namespace BuildXL.Cache.ContentStore.Service.Grpc
     /// </summary>
     public sealed class GrpcCopyClient : IShutdown<BoolResult>
     {
+        // Link to DistributedContentStoreSettings.MaxConcurrentCopyOperations
+        private const int _maxClientCount = 512;
+
+        private static readonly object lockObject = new object();
+        internal static readonly ConcurrentDictionary<(string, int, bool), GrpcCopyClient> _clientDict = new ConcurrentDictionary<(string, int, bool), GrpcCopyClient>(Environment.ProcessorCount, _maxClientCount);
+        private static Task _backgroundCleaningTask;
+        private static CancellationTokenSource _backgroundCleaningTaskTokenSource;
+
         private readonly Channel _channel;
         private readonly ContentServer.ContentServerClient _client;
         private readonly string _host;
         private readonly int _grpcPort;
         private bool _useCompression;
+        internal DateTime _lastUseTime;
+        internal int _uses;
+
+        static GrpcCopyClient()
+        {
+            RestartBackgroundCleanup();
+        }
+
+        private async static Task BackgroundCleanupAsync(CancellationTokenSource cts)
+        {
+            _backgroundCleaningTaskTokenSource = cts;
+            var ct = _backgroundCleaningTaskTokenSource.Token;
+
+            while (!ct.IsCancellationRequested)
+            {
+                if (!_clientDict.IsEmpty)
+                {
+                    var oneHourAgo = DateTime.UtcNow - TimeSpan.FromHours(1);
+
+                    foreach (var kvp in _clientDict)
+                    {
+                        var client = kvp.Value;
+                        if (client._uses == 0 && client._lastUseTime < oneHourAgo)
+                        {
+                            lock (lockObject)
+                            {
+                                if (_clientDict.TryRemove((client._host, client._grpcPort, client._useCompression), out GrpcCopyClient removedClient))
+                                {
+                                    // Cannot await within a lock
+                                    client._channel.ShutdownAsync().Start();

```

Comment:
>ShutdownAsync [](start = 52, length = 13)

Does the channel guarantee that the existing message is completed on it before it kills it?

Reply:
It does not guarantee anything. The assumption here is that the client will only have 0 uses if it is no longer in use.

---
In reply to: [289192667](https://github.com/microsoft/BuildXL/pull/391#discussion_r289192667) [](ancestors = 289192667)